http-gateways/SUBDOMAIN_GATEWAY.md

```diff
@@ -0,0 +1,271 @@
+    - [`X-Forwarded-Proto` (request header)](#x-forwarded-proto-request-header)
+    - [`X-Forwarded-Host` (request header)](#x-forwarded-host-request-header)
+  - [Request Query Parameters](#request-query-parameters)
+    - [`uri` (request query parameter)](#uri-request-query-parameter)
+- [HTTP Response](#http-response)
+  - [Response Headers](#response-headers)
+    - [`Location` (response header)](#location-response-header)
+      - [Use in interop with Path Gateway](#use-in-interop-with-path-gateway)
+      - [Use in URI router](#use-in-uri-router)
+- [Appendix: notes for implementers](#appendix-notes-for-implementers)
+  - [Migrating from Path to Subdomain Gateway](#migrating-from-path-to-subdomain-gateway)
+  - [DNS label limits](#dns-label-limits)
+  - [Security considerations](#security-considerations)
+  - [URI router](#uri-router)
+
+# HTTP API
+
+The API is a superset of [PATH_GATEWAY.md](./PATH_GATEWAY.md), the differences
+are documented below.
+
+The main one is that Subdomain Gateway expects CID to be present in the `Host` header.
+
+## `GET /[{path}][?{params}]`
+
+Downloads data at specified content path.
+
+- `path` – optional path to a file or a directory under the content root sent in `Host` HTTP header
+
+## `HEAD /[{path}][?{params}]`
+
+Same as GET, but does not return any payload.
+
+# HTTP Request
+
+Below MUST be implemented **in addition** to the [HTTP Request section from `PATH_GATEWAY.md`](./PATH_GATEWAY.md#http-request).
+
+## Request Headers
+
+### `Host` (request header)
+
+Defines the root that should be prepended to the `path` before IPFS content
+path resolution is performed.
+
+The value in `Host` header must be a valid FQDN with at least three DNS labels:
+a case-insensitive content root identifier followed by `ipfs` or `ipns`
+namespace, and finally the domain name used by the gateway.
+
+Converting `Host` into a content path depends on the nature of requested resource:
+
+- For content at `/ipfs/{cid}`:
+    - `Host: {cid-mbase32}.ipfs.example.net`
+        - Example: `Host: bafybeigdyrzt5sfp7udm7hu76uh7y26nf3efuylqabf3oclgtqy55fbzdi.ipfs.dweb.link`
+- For content at `/ipns/{libp2p-key}`:
+    - `Host: {libp2p-key-mbase36}.ipns.example.net`
+        - Example: `Host: k2k4r8jl0yz8qjgqbmc2cdu5hkqek5rj6flgnlkyywynci20j0iuyfuj.ipns.dweb.link`
+        - Note: Base36 must be used to ensure CIDv1 with ED25519 fits in a single DNS label (63 characters).
+- For content at `/ipns/{dnslink-name}`:
+    - `Host: {inlined-dnslink-name}.ipns.example.net`
+        - DNSLink names include `.` which means they  MUST be inlined into a single DNS label to provide unique origin and work with wildcard TLS certificates.
+            - DNSLink label encoding:
+                - Every `-` is replaced with `--`
+                - Every `.` is replaced with `-`
+            - DNSLink label decoding
+                - Every standalone `-` is replaced with `.`
+                - Every remaining `--` is replaced with `-`
```

Comment:
How does this account for `---`?

Reply:
Odd number of `-` here  does not look like a valid encoding:

- `/ipns/foo-bar.example.com` → `foo--bar-example-com.ipns.dweb.link`
- `/ipns/foo--bar.example.com` → `foo----bar-example-com.ipns.dweb.link` 
- `/ipns/foo---bar.example.com` → `foo------bar-example-com.ipns.dweb.link`


Is someone requests `foo---bar-example-com.ipns.dweb.link` the first `--` will  be turned into single `-` producing  `foo--bar.example.com`.  Content will load only if such domain name exists and DNSLink exists.

Reply:
Are `foo-.bar.example.com` or `foo.-bar.example.com` valid DNS domains? Those would both translate to `foo---bar.example.com`, right?

Reply:
no, `-` is not allowed at the beginning, nor end of a DNS label:

> The labels must follow the rules for ARPANET host names.  **They must start with a letter, end with a letter or digit**, and have as interior  characters only letters, digits, and hyphen.  There are also some restrictions on the length.  Labels must be 63 characters or less.
> – https://datatracker.ietf.org/doc/html/rfc1035#section-2.3.1